python/dolfinx/fem/petsc.py

```diff
@@ -174,47 +174,48 @@ def create_matrix_nest(a: typing.List[typing.List[Form]]) -> PETSc.Mat:
 
 # -- Vector assembly ---------------------------------------------------------
 
+
 @functools.singledispatch
-def assemble_vector(L: typing.Any, constants=None, coeffs=None) -> PETSc.Vec:
-    """Assemble linear form into a new PETSc vector.
+def assemble_vector(b: PETSc.Vec, L: Form, constants=None, coeffs=None) -> PETSc.Vec:
+    """Assemble linear form into an existing PETSc vector.
 
     Note:
-        The returned vector is not finalised, i.e. ghost values are not
-        accumulated on the owning processes.
+        The vector is not zeroed before assembly and it is not
+        finalised, i.e. ghost values are not accumulated on the owning
+        processes.
 
     Args:
-        L: A linear form.
+        b: Vector to assemble the contribution of the linear form into.
+        L: A linear form to assemble into ``b``.
 
     Returns:
         An assembled vector.
 
     """
-    b = create_petsc_vector(L.function_spaces[0].dofmap.index_map,
-                            L.function_spaces[0].dofmap.index_map_bs)
     with b.localForm() as b_local:
-        _assemble._assemble_vector_array(b_local.array_w, L, constants, coeffs)
+        _assemble.assemble_vector(b_local.array_w, L, constants, coeffs)
     return b
 
 
-@assemble_vector.register(PETSc.Vec)
-def _assemble_vector_vec(b: PETSc.Vec, L: Form, constants=None, coeffs=None) -> PETSc.Vec:
-    """Assemble linear form into an existing PETSc vector.
+@assemble_vector.register
+def _assemble_vector_new(L: Form, constants=None, coeffs=None) -> PETSc.Vec:
+    """Assemble linear form into a new PETSc vector.
```

Comment:
At first glance I thought this meant PETSc had a new version of vector. How about this instead?

```suggestion
    """Create a PETSc vector and assemble linear form into it.
```